### ISSUE TITLE
SSL: CERTIFICATE_VERIFY_FAILED

### ISSUE THREAD

**Issue by emsecrist** [boto/botocore#1553]
I have tried various approaches to resolve an SSL:CERTIFICATE_VERIFY_FAILED error when attempting to start an AWS Glacier Job. After upgrading to boto3 1.9.1, I received the following error, which is more specific than the error from previous versions of boto3:

botocore.exceptions.SSLError: SSL validation failed for https://glacier.us-west-2.amazonaws.com/-/vaults/xxx-vault/jobs [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:600)

I think the "-" after ".com" and before "vaults" in the URL looks incorrect, but I am uncertain. Any help here would be much appreciated.

**Comment by joguSD** [comment]
The dash is correct (it tells glacier to autofill the account id), and in general the path of the URL should be unrelated to SSL certification validation. 


* Could you provide a minimal example of the code you're using to produce this error?
* Could you provide sanitized debug logs by adding: `boto3.set_stream_logger('')`.
* What platform are you on? 
* Has this code worked in the past?

**Comment by joguSD** [comment]
A quick look at the debug logs looks like you're loading a custom CA bundle:
```
2018-09-11 15:14:02,233 botocore.session [DEBUG] Loading variable ca_bundle from config file with value '/etc/pki/tls/cert.pem'.
```
It's possible that glacier is providing a cert on some requests that isn't trusted by the bundle you're providing. Though, I'm not entirely certain the best way to go about determining which certificate is causing the problem as we don't log that anywhere.

Just for clarity, you're not using a proxy correct?

**Comment by emsecrist** [comment]
The CA bundle you listed is the same bundle that was used previously. Why would that bundle work sometimes, but not at other times/

I have a Direct Connect connection to the us-west-2 region, so I would assume I am not using a proxy. Unless a proxy can be configured with Direct Connect?

**Comment by joguSD** [comment]
Out of curiosity, is there any reason in particular that you're setting the CA bundle explicitly?

I did notice that in the debug logs the connection was being reset fairly often meaning a reconnect at thus TLS is renegotiated. My only guess is that depending on the glacier host you connect to maybe sometimes you get different certificates, some you trust and some you don't. This is purely speculation, though.

**Comment by emsecrist** [comment]
Actually I explicitly defined the CA bundle as part of troubleshooting this and forgot that it was still in place. I removed the explicit definition and am now using the default bundle, but this error still occurs when I run the code.

**Comment by joguSD** [comment]
Could you post the debug logs using the default certificate?
Also, were you running into this problem before upgrading or only after upgrading?

**Comment by emsecrist** [comment]
I will post the debug logs using the default certificate tomorrow morning
(9/12) when I get back to the office. I was running into this problem
before upgrading boto3. I upgraded boto3 to try and fix this problem.

**Comment by shubhamsawantsjsu** [comment]
I have got the solution for this.

Don't install boto3 using pip/pip3.

Use following steps::

$ git clone https://github.com/boto/boto3.git
$ cd boto3
$ virtualenv venv
...
$ . venv/bin/activate
$ pip install -r requirements.txt
$ pip install -e .

It worked for me. Initially I was facing the same SSL Validation Exception problem. Then I uninstalled the boto3 (pip3 uninstall boto3) and re-installed it using virtualenv as stated above. Now, it is working fine.
